biodatacatalyst-ui/src/main/webapp/picsureui/search-interface/tour-view.js

```diff
@@ -23,26 +23,81 @@ define([
         return loadingScreen;
     }
     var TourView = BB.View.extend({
-        initialize: function(){
+        initialize: function(opts){
+            this.opts = opts;
             keyboardNav.addNavigableView('chardinJs', this);
             this.on({
-				'keynav-arrowup document': this.prevStep.bind(this),
-				'keynav-arrowdown document': this.nextStep.bind(this),
-				'keynav-arrowright document': this.nextStep.bind(this),
-				'keynav-arrowleft document': this.prevStep.bind(this),
-				'keynav-enter': this.nextStep.bind(this),
-				'keynav-space': this.nextStep.bind(this),
-                'keynav-escape': this.stopTour.bind(this),
-			});
+              'keynav-arrowup document': this.prevStep.bind(this),
+              'keynav-arrowdown document': this.nextStep.bind(this),
+              'keynav-arrowright document': this.nextStep.bind(this),
+              'keynav-arrowleft document': this.prevStep.bind(this),
+              'keynav-enter': this.nextStep.bind(this),
+              'keynav-space': this.nextStep.bind(this),
+              'keynav-escape': this.stopTour.bind(this),
+            });
+            Backbone.pubSub.on('searchResultsRenderCompleted', this.initChardinJs.bind(this));
         },
         initChardinJs: function() {
-            this.loadingScreen = showFullScreenLoadingView($);
-            //Artificial delay to allow the set up render finish rendering. chardin seems to be slow to find new elements.
-			delay(2750).then(()=> { //TODO: find a better way to do this
-                this.overlay =  $('body').chardinJs({ url: './search-interface/guide-me.json'})
-                delay(500).then(() => this.startTour()); //Delay again because chardin is slow to get elements
-         });
+            let callback = _.once(() => {
+                this.checkIfPreRenderCompleted(this.opts.idsToWaitFor).then(() => {
+                    this.checkIfOverlayIsReady().then(() => {
+                        this.startTour();
+                    }).catch((err) => {
+                        console.error(err);
+                        this.loadingScreen.remove();
+                        alert('Something went wrong. Please try again.');
+                    });
+                }).catch((err) => {
+                    console.error(err);
+                    this.loadingScreen.remove();
+                    alert('Something went wrong. Please try again.');
+                });
+            });
+            $(document).ready(callback);
 		},
+        checkIfPreRenderCompleted: function(idsToWaitFor) {
+            return new Promise(function(resolve, reject) {
+                const timeoutCount = 10;
+                let count = 0;
+                let elementsToWaitFor = [];
+                idsToWaitFor?.forEach(elementId => {
+                    elementsToWaitFor.push(document.getElementById(elementId));
+                });
```

Comment:
```suggestion
        checkIfPreRenderCompleted: function(idsToWaitFor = []) {
            return new Promise(function(resolve, reject) {
                const timeoutCount = 10;
                let count = 0;
                const elementsToWaitFor = idsToWaitFor.map(elementId => document.getElementById(elementId));
```
There's nothing wrong with what you have already, just an optional suggestion (if it works).

Reply:
Looks cleaner to me! Thanks